content/rs/developing/crdbs/developing-streams-crdb.md

```diff
@@ -115,158 +123,155 @@ Using XREAD to iterate through stream entries is a popular pattern (Very similar
-Cross-region replication is slow and bandwidth is limited so we chose to go with a different approach:
+In consumer groups, every XREADGROUP and every XACK can change the state of a group or consumer.
+but this cross-location replication is slow and bandwidth is limited.
+To maintain consumer groups in Active-Active databases with optimal performance:
 
-1. Group existence (CREATE/DESTROY) is replicated
-1. XACk is replicated, but not every XACK (will be explained later)
-1. Everything else is local (XGROUP SETID, DELCONSUMER , etc.)
+1. Group existence (CREATE/DESTROY) is replicated.
+1. Most XACK operations are replicated.
+1. Other operations are local, such as: XGROUP, SETID, DELCONSUMER
 
 For example:
 
-| time | Region 1              | Region 2              |
-| ---- | --------------------- | --------------------- |
-| t1 |XADD x 110-0 f1 v1 |
-| t2 | XGROUP CREATE x g1 0 | 
-| t3 | XREADGROUP GROUP g1 Alice STREAMS x > → [110-0] | 
-| t4 | sync
-| t5 | XRANGE x - + → [110-0] | XRANGE x - + → [110-0]
-| t6 | XINFO GROUPS x -> [g1] | XINFO GROUPS x -> [g1]
-| t7 | XINFO CONSUMERS x -> [Alice] | XINFO CONSUMERS x -> []
-| t8 | XPENDING x g1 - + → [110-0] | XPENDING x g1 - + → []
-
-As you can see, Region 2 seems to be completely unaware of any group-related activity and it seems that redirecting the XREADGROUP traffic from Region 1 to Region 2 will result in reading the same entries that were already read.
+| time | Location 1                                        | Location 2               |
+| ---- | ------------------------------------------------- | ------------------------ |
+| t1   | XADD x 110-0 f1 v1                                |                          |
+| t2   | XGROUP CREATE x g1 0                              |                          |
+| t3   | XREADGROUP GROUP g1 Alice STREAMS x > --> [110-0] |                          |
+| t4   | --- sync ---                                      | --- sync ---             |
+| t5   | XRANGE x - + --> [110-0]                          | XRANGE x - + --> [110-0] |
+| t6   | XINFO GROUPS x -> [g1]                            | XINFO GROUPS x -> [g1]   |
+| t7   | XINFO CONSUMERS x -> [Alice]                      | XINFO CONSUMERS x -> []  |
+| t8   | XPENDING x g1 - + --> [110-0]                     | XPENDING x g1 - + --> [] |
+
+As you can see, Location 2 seems to be completely unaware of any group-related activity and it seems that redirecting the XREADGROUP traffic from Location 1 to Location 2 will result in reading the same entries that were already read.
 Given that the usage pattern is standard (Every entry that was read using XREADGROUP is acknowledged with XACK after processing it) we have a way to tackle this issue.
 
 Note that using XREADGROUP across instances can result in instances reading the same entries because CRDT Streams is designed to handle “at least once” reads (or have only one reader at a time).
```

Comment:
@guybe7 What did you mean by "we have a way to tackle this issue"?

Reply:
the issue is the fact the Location 2 is unaware of any group-related activity in Location 1. it poses this problem: "redirecting the XREADGROUP traffic from Location 1 to Location 2 will result in reading the same entries that were already read"
so the issue leads to a problem and the solution is described in the section "### XREADGROUP redirection"

(i thin that the paragraph "	Note that using XREADGROUP across instances.." was not there originally so the transition from "here is a problem" to "the section that describes the solution" was smoother)